ranger/config/commands.py

```diff
@@ -1176,43 +1177,203 @@ def execute(self):
+        return lambda rwx_str: (
+            (r_flag if 'r' in rwx_str else 0) |
+            (w_flag if 'w' in rwx_str else 0) |
+            (x_flag if 'x' in rwx_str else 0))
+
+    def get_mode_change_funcs(self, who, operator, perms):
+        '''returns mode-change-functions (mcf) for given action.'''
+
+        if perms == '' and operator in '+-':
+            return tuple()
+        mcfs, affected_bits = [], 0
+        if who == '':
+            clear_bits = self.MASKS['a']['rwxst']
+            affected_bits = clear_bits & ~self.get_umask()
+        else:
+            for target in who:
+                affected_bits |= self.MASKS[target]['rwxst']
+            clear_bits = affected_bits
+
+        if operator == '=':
+            mcfs.append(lambda omode, lmode: lmode & ~clear_bits)
+
+        if len(perms) == 1  and  perms in 'ugo':    # permcopy
+            source_rwx_func = self.fg_extract_rwx_for_source(perms)
+            target_mask_func = self.fg_get_mode_for_target('a')
+            mode2affected = lambda mode: (
+                affected_bits & target_mask_func(source_rwx_func(mode)))
+            mcfs.append(
+                (lambda omode, lmode: lmode | mode2affected(omode)) if
+                  operator in '+=' else
+                (lambda omode, lmode: lmode & ~(mode2affected(omode))))
+        else:
+            mask, a_x = 0, self.MASKS['a']['x']
+            if operator == '+'  and  'X' in perms:
+                mcfs.append(lambda omode, lmode: (
+                    lmode | (a_x & affected_bits) if stat.S_ISDIR(omode) or
+                        (omode & a_x) else lmode))
+            perms = re.sub('X', '', perms)
+            for perm in perms:
+                if who == '':
+                    mask |= self.MASKS['a'][perm]
+                else:
+                    for target in who:
+                        mask |= self.MASKS[target][perm]
+            aff_mask = mask & affected_bits
+            mcfs.append(
+                (lambda omode, lmode: lmode | aff_mask) if operator in '+=' else
+                (lambda omode, lmode: lmode & ~aff_mask))
+
+        return mcfs
+
+    def parse_mode(self, mode_str):
+        '''
+        parse given mode.
+
+        returns a list of mode-change-functions (mcf) with each mcf
+            mcf(orig_mode, last_computed_mode) -> new_computed_mode
+            or
+            mcf(orig_mode, last_computed_mode) -> new_orig_mode, new_computed_mode
+        '''
+        mcfs = []
+        if re.match(self.OCTAL_MODE_RE, mode_str):
+            try:
+                mode = int(mode_str, 8)
+                if mode < 0  or mode > 0o7777:
+                    raise ValueError()
+                mcfs.append(lambda omode, lmode: mode)
+            except ValueError:
+                self.fm.notify(
+                    "Need an octal number between 0o0 and 0o7777!", bad=True)
+            return mcfs
+
+        set_omode = lambda omode, lmode: (lmode | (omode & ~0o7777), lmode)
```

Comment:
Isn't `omode & ~0o7777` the opposite of a tautology, i.e., always false? Since `~0o7777` is `0`?

Reply:
Also, I'm unclear about why we're changing the omode here?

Reply:
No `~0o7777` is not `0`.  Why should it be? More than 12 bits are used for the mode bits (in a typical OS). And here we have one of the special cases: the wish "all of the modes" includes changes of the modes, that depend on the current original mode "omode", e.g. a mode-spec using `X` has to test if the current file under investigation is a directory. This test is done by checking the bit `0o040000` in `omode` (see; there are more than 12 bits!). And to check this I have to pass-through the `omode` (the original mode) to the mode-chaning functions in order to implement e.g. the `X`-modespec. 

Why do I need to change the `omode` during the operations? Because "original mode" means: "mode as it was seen at the beginning of the current 'clause'", so take e.g. `chmod g=u,o=g ./file`. That has two clauses and for the second one `o=g` the `omode` is the mode after `g=u` was computed!  Hence I have to keep track of the `omode` and I have to keep the higher bits (above 12) in order for the following mode-specs (that may depend on the higher bits, see `X` example) to do the right things. Hence `omode & ~0o7777` means: keep all the higher bits.

Yes, there are a lot of special cases.  (One can have  `chmod g=u-w ./file`  and also `chmod =w=rx ./file`, etc.)